Add signals for retrieving additional data from solver_cg

simonsticko opened the pull request:
SolverCG can estimate coefficients, eigenvalues and condition numbers during the
solution process and output these to deallog. Add boost::signals2 to be able to
communicate these estimates back to the caller. Add one signal for each
quantity. Use the same signal for condition number and all condition numbers.

tjhei commented on include/deal.II/lac/solver_cg.h:
```diff
@@ -166,6 +166,24 @@ class SolverCG : public Solver<VECTOR>
 
+  /**
+   * Connect a slot to retrieve the CG coefficients.
+   */
```
please document what the two arguments of the ```slot``` are going to be.

bangerth replied:
also how often it will be called (only once at the end, or once per iteration?). same for the one below.

tjhei commented:
Very nice!

Right now you need to enable the computations in AdditionalData and attach slots for them to work. I think we should do one of the following:
  1. deprecate the options in AdditionalData and compute if a slot is attached.
  2. enable appropriate flags if a slot is attached.
  3. report an error if you attach a slot and you forgot to enable the setting in AdditionalData.

We also need a unit test to check that this works. Do you know how to add one?

tjhei commented on include/deal.II/lac/solver_cg.h:
```diff
@@ -166,6 +166,24 @@ class SolverCG : public Solver<VECTOR>
+  connect_condition_number_slot(const std_cxx11::function<void (double)> &slot);
+
+  /**
+   * Connect a slot to retrieve the estimated eigenvalues.
```
document the argument of the slot (contents, length)

kronbichler commented:
Once we identified a strategy as listed by @tjhei, we should apply the same to GMRES which can also compute eigenvalue estimates. My favorite would be option 1, i.e., just have one option to get at this info and deprecate the old version. Maintaining several options for the same thing is almost always a pain.

simonsticko commented:
Okey, it seems like alternative 1 is the preferred solution. I'll make a few changes. What is appropriate way to create a test? To write a new one or to add some extra stuff to an existing test?

tjhei commented:
> What is appropriate way to create a test? To write a new one or to add some extra stuff to an existing test?

Create a new test by copying and adapting one or more existing tests. Some tests you could look at:
lapack/solver_cg.cc (checks the condition estimate stuff)
tests/lac/solver_monitor.cc (an example for testing slots)

bangerth commented:
@simonsticko : Information about the testsuite is at http://dealii.org/developer/developers/testsuite.html

As for the three options @tjhei listed: My suggestion would be to just ignore all of the stuff in `AdditionalData`. Let's just create a separate pathway, i.e., just call your new signal whether the flag is set or not. We can then at a later stage (or if you'd like as part of the patch) deprecate the flag in the `AdditionalData` struct that duplicates the signal mechanism. So I think you're mostly there already.

As for the addition of similar functionality to GMRES: if you want to work on that, that would of course be awesome, but that's not necessary to get this one patch here accepted.

Other than that, many thanks for working on this -- this is pretty nice. Please let us know if you have difficulties and need help!

bangerth commented:
@simonsticko : Any news on this?

simonsticko commented:
I have not forgotten :) Just busy with a lot of other things. I have one question however. If additional data is deprecated and removed in some future release, will that not force anyone wanting to use this to switch to c++11? Maybe this is not a problem, but I got the impression that this wanted to be avoided.

bangerth commented:
Your code doesn't use C++11, so everything should continue to work fine. Or do I miss something?

simonsticko commented:
Ahh, sorry, I thought std_cxx11::bind required c++11, but I see now that it doesn't.

simonsticko commented:
Sorry for the long time between updates. A few issues:

1.
Right now the condition number in gmres is calculated by max(|eig|)/min(|eig|) as @kronbichler suggested here:
https://groups.google.com/forum/#!searchin/dealii/condition$20number/dealii/wMNlxxOkMn4/3dMejbGyd54J
Does this estimate hold also in the case where the matrix is not normal? My NLA is rusty.

2.
The only way I could deprecate the old use of AdditionalData was to add a new 
constructor and mark the old one as DEAL_II_DEPRECATED. Is there a better way to do this?

3.
Right now the test solver_signals.cc connects all possible slots and writes the 
output to log. This tests that all slots can be connected, but is this what I want to test?

kronbichler commented:
Regarding the condition number: @simonsticko is right that max(eig)/min(eig) only holds for normal matrices. I have not checked how other people/software do and know too little about this subject to make a qualified statement. I would assume that the mathematically correct approach would be to compute singular values of the Hessenberg matrix instead of the eigenvalues. However, the question would be how far away from the actual condition number one can get: Keep in mind that the Hessenberg matrix and any information you extract from there is an approximation of the operator along the Krylov subspace `r0, A*r0,A^2*r0,...`, so it is an estimate already and it might be as far off as the eigenvalue bound vs. real condition number on a non-normal matrix.

That said, my preference would be to do things _less wrong_ and compute the svd in case the condition number was requested (`LAPACKFullMatrix::compute_svd` and `LAPACKFullMatrix::singular_value(i)`). These are small matrices in any case, so in the worst case one ends up with two O(n^3) operations instead of one with `n` at most a few hundreds (28 with default settings).

Regarding DEAL_II_DEPRECATED: It does look good to me this way. The same for the tests, but I'll let @bangerth have a comment on the signal part.

bangerth commented on include/deal.II/lac/solver_cg.h:
```diff
@@ -96,6 +96,15 @@ class SolverCG : public Solver<VECTOR>
   typedef types::global_dof_index size_type;
 
+  template<class T>
+  using signal=boost::signals2::signal<T>;
```
This is a use of C++11 (template using statements didn't exist before). Please just spell out the fully qualified name wherever you use it.

bangerth commented on include/deal.II/lac/solver_cg.h:
```diff
@@ -166,6 +182,33 @@ class SolverCG : public Solver<VECTOR>
+  /**
+   * Connect a slot to retrieve the CG coefficients. The slot will be called
+   * with alpha as the first argument and with beta as the second argument.
+   * Called once per iteration.
```
Can you say what alpha and beta refer to here?

bangerth replied:
For example by pointing to a place where a pseudo-algorithm is listed that uses these names.

bangerth commented on include/deal.II/lac/solver_cg.h:
```diff
@@ -166,6 +182,33 @@ class SolverCG : public Solver<VECTOR>
+  /**
+   * Connect a slot to retrieve the estimated condition number.
+   * Called on each iteration if every_iteration=true otherwise called
+   * after all iterations.
```
"after all iterations" is misunderstandable. How about "otherwise called at the end of iterations once convergence has been achieved"?

simonsticko replied:
Right now I think the signals will be called even if the solver didn't converge, since SolverControl::success is checked after the call to compute_eigs_and_cond. Do you think it would make more sense not to call the signals if the solver diverged?

bangerth replied:
Good point. How about this as a replacement text: "...otherwise called once when iterations are ended (i.e., either because convergence has been achieved, or because divergence has been detected)."?

bangerth commented on include/deal.II/lac/solver_cg.h:
```diff
@@ -166,6 +182,33 @@ class SolverCG : public Solver<VECTOR>
+   */
+  boost::signals2::connection
+  connect_eigenvalues_slot(const std_cxx11::function<EigenvalueSlot> &slot,
+                           const bool every_iteration=false);
```
Since you spell out the type of the function everywhere else, I would avoid the typedef `Eigenvalue` slot here as well. This also makes it clear immediately (without having to look up the function signature) what the function arguments are.

Same with "after all iterations" as above.

bangerth commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -160,6 +161,15 @@ template <class VECTOR = Vector<double> >
 public:
+
+  template<class T>
+  using signal=boost::signals2::signal<T>;
```
same as above.

bangerth commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -610,10 +769,16 @@ SolverGMRES<VECTOR>::solve (const MATRIX         &A,
   // restart
   unsigned int accumulated_iterations = 0;
 
+  bool do_eigenvalues=
```
mark the `bool` as `const` since it's never going to change.

bangerth commented:
Great, I like his a lot! I only have a few stylistic issues that I've marked up in the patch.

As for the question of what these functions compute: My linear algebra-fu is not strong enough to really tell whether the computed values make sense or not. I think the solution to these things is generally to just add a section to the general class documentation that states in explicit terms what it is that is being computed. Users can then compare this with what it is they want. In other words, just add a disclaimer that what is being computed is X and may only apply if the matrix has property Y. If someone has a matrix that does not have property Y, then they will have to decide whether this the things your patch does is appropriate for them or not.

simonsticko commented:
Fixed the stylistic issues.

Changed the computation of eigenvalues to using svd instead, as kronbichler described.
I didn't manage to do this without copying the Hessenberg matrix an extra time, since
compute_svd changes the state of the matrix. I don't know if this can be avoided somehow?

Added a disclaimer in gmres about how the eigenvalues and condition number are computed.

Updated the documentation under the header "Eigenvalue computation" in solver_cg
to better correspond to the signals way.

bangerth commented:
Nice work!

I have looked over the code and am ok with it. @kronbichler -- you seemed to understand the linear algebra side better. Can you take a look at the two functions that compute the eigenvalues? If you're ok with these two, please merge it all.

kronbichler commented:
This is very nice. I like it a lot. The description looks good and the algorithms are as we discussed.

Regarding copying the Hessenberg matrix: That's exactly what I would have done, so that's fine.

kronbichler commented:
@simonsticko We have a problem with the tester: http://cdash.kyomu.43-1.org/testDetails.php?test=16540806&build=9966
The problem is that the eigenvalue estimates from the GMRES solver are not sorted (they are in turn the result of the LAPACK function `dgeev` which does not specify a particular order for eigenvalues; in the general complex case, there is of course no straight-forward ordering).
Would you mind adding some simple sort mechanism (similarly to `internal::SolverGMRES::complex_less_pred`) for the test?

simonsticko commented:
Can I do this from this feature branch or do I need to create a new branch and a new pull request?

bangerth commented:
You'll need to create a new branch and pull request.
